File: frame_karaoke/karaoke_app.py
"""Main karaoke application for Frame glasses."""

import asyncio
import logging
import os
from typing import Optional
from frame_sdk import Frame
from frame_sdk.display import Alignment

from .models import AppState, CurrentSong
from .services import ACRCloudService, LRCService
from .managers import (
    RecognitionManager,
    LyricsManager,
    PositionTracker,
    DisplayManager,
    HistoryManager,
)

logger = logging.getLogger(__name__)


class KaraokeApp:
    """
    Main karaoke application.

    Coordinates audio capture, song recognition, lyrics display,
    and all other karaoke functionality for Frame glasses.
    """

    def __init__(
        self,
        acrcloud_host: str,
        acrcloud_access_key: str,
        acrcloud_secret_key: str,
        frame_address: Optional[str] = None
    ):
        """
        Initialize karaoke app.

        Args:
            acrcloud_host: ACRCloud API host
            acrcloud_access_key: ACRCloud access key
            acrcloud_secret_key: ACRCloud secret key
            frame_address: Frame pairing code (None = connect to any Frame)
        """
        self.frame_address = frame_address
        self.frame: Optional[Frame] = None

        # Services
        self.acrcloud_service = ACRCloudService(
            acrcloud_host,
            acrcloud_access_key,
            acrcloud_secret_key
        )
        self.lrc_service = LRCService()

        # Managers
        self.recognition_manager = RecognitionManager(self.acrcloud_service)
        self.lyrics_manager = LyricsManager(self.lrc_service)
        self.position_tracker = PositionTracker()
        self.display_manager = DisplayManager()
        self.history_manager = HistoryManager()

        # State
        self.current_song: Optional[CurrentSong] = None
        self.app_state: AppState = AppState.LISTENING
        self.running: bool = False

    async def start(self):
        """Start the karaoke application."""
        logger.info("Starting Frame Karaoke...")

        try:
            async with Frame(self.frame_address) as frame:
                self.frame = frame
                self.running = True

                logger.info("Connected to Frame glasses")

                # Show welcome message
                await self._update_display()

                # Start recognition
                self.recognition_manager.start_listening()

                # Run main loops concurrently
                await asyncio.gather(
                    self._audio_capture_loop(),
                    self._recognition_loop(),
                    self._display_loop(),
                )

        except KeyboardInterrupt:
            logger.info("Received interrupt signal")
        except Exception as e:
            logger.error(f"Application error: {e}", exc_info=True)
        finally:
            await self.stop()

    async def stop(self):
        """Stop the karaoke application."""
        logger.info("Stopping Frame Karaoke...")
        self.running = False
        self.recognition_manager.stop_listening()

    async def _audio_capture_loop(self):
        """Continuously capture audio from Frame microphone."""
        logger.info("Starting audio capture loop")

        while self.running:
            try:
                # Record 2-second chunks for continuous monitoring
                audio_data = await self.frame.microphone.record_audio(
                    max_length_in_seconds=2
                )

                # Add to recognition buffer
                self.recognition_manager.add_audio_chunk(audio_data)

            except Exception as e:
                logger.error(f"Audio capture error: {e}")
                await asyncio.sleep(1)  # Prevent tight error loop

    async def _recognition_loop(self):
        """Periodically perform song recognition."""
        logger.info("Starting recognition loop")

        while self.running:
            try:
                # Check if it's time to recognize
                if self.recognition_manager.should_recognize():
                    await self._perform_recognition()

                await asyncio.sleep(1)  # Check every second

            except Exception as e:
                logger.error(f"Recognition loop error: {e}")
                await asyncio.sleep(5)

    async def _display_loop(self):
        """Periodically update Frame display."""
        logger.info("Starting display loop")

        while self.running:
            try:
                await self._update_display()
                await asyncio.sleep(0.5)  # Update every 500ms

            except Exception as e:
                logger.error(f"Display loop error: {e}")
                await asyncio.sleep(1)

    async def _perform_recognition(self):
        """Perform song recognition and handle result."""
        result = await self.recognition_manager.recognize()

        if not result or result.error:
            # No song detected or error
            if self.current_song:
                # Check if song has ended
                position = self.position_tracker.get_current_position()
                if position > self.current_song.duration + 5:
                    logger.info("Song appears to have ended")
                    await self._end_song()
            return

        # Got a recognition result
        await self._handle_recognition_result(result)

    async def _handle_recognition_result(self, result):
        """
        Handle song recognition result.

        Args:
            result: RecognitionResult from ACRCloud
        """
        # Check if this is a new song or continuation
        if self.current_song is None:
            # New song detected
            await self._start_new_song(result)

        else:
            # Check if same song or different song
            if (result.title == self.current_song.title and
                result.artist == self.current_song.artist):
                # Same song - update position
                await self._update_song_position(result)
            else:
                # Different song detected
                if result.confidence > 0.75:  # High confidence for song switch
                    await self._switch_song(result)

    async def _start_new_song(self, result):
        """
        Start tracking a new song.

        Args:
            result: RecognitionResult
        """
        logger.info(f"Starting new song: '{result.title}' by '{result.artist}'")

        # Create CurrentSong
        self.current_song = CurrentSong(
            title=result.title,
            artist=result.artist,
            album=result.album,
            duration=result.duration or 180.0,  # Default 3 minutes
            detected_at=asyncio.get_event_loop().time(),
            has_lyrics=False,
            confidence=result.confidence
        )

        # Start position tracking
        if result.offset_seconds is not None:
            self.position_tracker.start_song(
                asyncio.get_event_loop().time(),
                result.offset_seconds,
                getattr(result, 'api_latency', 0)
            )

        # Add to history
        self.history_manager.add_song(self.current_song)

        # Set state to processing while we fetch lyrics
        self.app_state = AppState.PROCESSING

        # Fetch lyrics
        lrc_data = await self.lyrics_manager.fetch_lyrics(self.current_song)

        if lrc_data:
            self.current_song.has_lyrics = True
            self.current_song.lrc_data = lrc_data
            self.app_state = AppState.SONG_DETECTED_WITH_LYRICS
            logger.info(f"Lyrics loaded: {len(lrc_data)} lines")
        else:
            self.app_state = AppState.SONG_DETECTED_NO_LYRICS
            logger.info("No lyrics available")

    async def _update_song_position(self, result):
        """
        Update position tracking for current song.

        Args:
            result: RecognitionResult
        """
        if result.offset_seconds is not None:
            self.position_tracker.recalibrate(
                result.offset_seconds,
                asyncio.get_event_loop().time(),
                result.confidence,
                getattr(result, 'api_latency', 0)
            )

    async def _switch_song(self, result):
        """
        Switch to a different song.

        Args:
            result: RecognitionResult for new song
        """
        logger.info(
            f"Switching song: '{self.current_song.title}' -> '{result.title}'"
        )
        await self._end_song()
        await self._start_new_song(result)

    async def _end_song(self):
        """End current song tracking."""
        if self.current_song:
            logger.info(f"Ending song: '{self.current_song.title}'")

        self.current_song = None
        self.app_state = AppState.LISTENING
        self.position_tracker.reset()

    async def _update_display(self):
        """Update Frame display with current state."""
        if not self.frame:
            return

        # Get current position
        position = self.position_tracker.get_current_position()

        # Get current and next lyrics chunks
        current_chunk = None
        next_chunk = None

        if self.current_song and self.current_song.has_lyrics:
            current_chunk = self.lyrics_manager.get_current_chunk(position)
            next_chunk = self.lyrics_manager.get_next_chunk(position)

        # Format display
        display_text = self.display_manager.format_display(
            self.app_state,
            self.current_song,
            current_chunk,
            next_chunk,
            position
        )

        # Send to Frame
        try:
            await self.frame.display.show_text(
                display_text,
                align=Alignment.TOP_LEFT
            )
        except Exception as e:
            logger.error(f"Display update error: {e}")

File: flutter_karaoke/lib/karaoke_app.dart
import 'dart:async';
import 'dart:typed_data';
import 'package:frame_sdk/frame_sdk.dart';
import 'package:frame_sdk/display.dart';
import 'package:record/record.dart';

import 'models/app_models.dart';
import 'services/acrcloud_service.dart';
import 'services/lrc_service.dart';
import 'managers/recognition_manager.dart';
import 'managers/lyrics_manager.dart';
import 'managers/position_tracker.dart';
import 'managers/display_manager.dart';
import 'managers/history_manager.dart';

/// Audio source for the app
enum AudioSource {
  frame,      // Frame glasses microphone
  phone,      // Phone's built-in microphone
}

/// Main karaoke application for Frame glasses
class KaraokeApp {
  final Frame? frame;  // Optional now for test mode
  final String acrcloudHost;
  final String acrcloudAccessKey;
  final String acrcloudSecretKey;
  final AudioSource audioSource;

  // Phone microphone recorder (for test mode)
  final AudioRecorder _audioRecorder = AudioRecorder();

  // Services
  late final ACRCloudService _acrcloudService;
  late final LRCService _lrcService;

  // Managers
  late final RecognitionManager _recognitionManager;
  late final LyricsManager _lyricsManager;
  late final PositionTracker _positionTracker;
  late final DisplayManager _displayManager;
  late final HistoryManager _historyManager;

  // State
  CurrentSong? currentSong;
  AppState appState = AppState.listening;
  bool _isRunning = false;

  // Timers
  Timer? _recognitionTimer;
  Timer? _displayTimer;
  Timer? _phoneAudioTimer;

  // Display callback for showing lyrics in app UI
  Function(String)? onDisplayUpdate;

  KaraokeApp({
    this.frame,
    required this.acrcloudHost,
    required this.acrcloudAccessKey,
    required this.acrcloudSecretKey,
    this.audioSource = AudioSource.frame,
    this.onDisplayUpdate,
  }) {
    _initializeServices();
  }

  void _initializeServices() {
    // Initialize services
    _acrcloudService = ACRCloudService(
      host: acrcloudHost,
      accessKey: acrcloudAccessKey,
      secretKey: acrcloudSecretKey,
    );
    _lrcService = LRCService();

    // Initialize managers
    _recognitionManager = RecognitionManager(
      acrcloudService: _acrcloudService,
    );
    _lyricsManager = LyricsManager(_lrcService);
    _positionTracker = PositionTracker();
    _displayManager = DisplayManager();
    _historyManager = HistoryManager();
  }

  /// Start the karaoke app
  Future<void> start() async {
    _isRunning = true;

    // Show welcome message
    await _updateDisplay();

    // Start recognition
    _recognitionManager.startListening();

    // Start audio capture based on source
    if (audioSource == AudioSource.phone) {
      await _startPhoneAudioCapture();
    }
    // For Frame, audio will be added via addAudioChunk() calls

    // Start periodic recognition checks (every 1 second)
    _recognitionTimer = Timer.periodic(const Duration(seconds: 1), (_) async {
      if (_recognitionManager.shouldRecognize()) {
        await _performRecognition();
      }
    });

    // Start display updates (every 500ms)
    _displayTimer = Timer.periodic(const Duration(milliseconds: 500), (_) async {
      if (_displayManager.shouldUpdate()) {
        await _updateDisplay();
      }
    });
  }

  /// Start phone microphone audio capture
  Future<void> _startPhoneAudioCapture() async {
    try {
      // Check and request permission
      if (!await _audioRecorder.hasPermission()) {
        print('Microphone permission denied');
        return;
      }

      // Start recording in chunks
      _phoneAudioTimer = Timer.periodic(const Duration(seconds: 2), (_) async {
        try {
          // Start recording to stream
          final stream = await _audioRecorder.startStream(
            const RecordConfig(
              encoder: AudioEncoder.pcm16bits,
              sampleRate: 16000,
              numChannels: 1,
            ),
          );

          // Collect 2 seconds of audio
          final chunks = <Uint8List>[];
          final subscription = stream.listen((data) {
            chunks.add(Uint8List.fromList(data));
          });

          // Wait 2 seconds
          await Future.delayed(const Duration(seconds: 2));

          // Stop recording
          await subscription.cancel();
          await _audioRecorder.stop();

          // Combine chunks and add to recognition buffer
          if (chunks.isNotEmpty) {
            final combined = Uint8List.fromList(
              chunks.expand((chunk) => chunk).toList(),
            );
            addAudioChunk(combined);
          }
        } catch (e) {
          print('Phone audio capture error: $e');
        }
      });
    } catch (e) {
      print('Failed to start phone audio: $e');
    }
  }

  /// Stop the karaoke app
  Future<void> stop() async {
    _isRunning = false;
    _recognitionManager.stopListening();
    _recognitionTimer?.cancel();
    _displayTimer?.cancel();
    _phoneAudioTimer?.cancel();

    // Stop phone audio recording
    if (audioSource == AudioSource.phone) {
      try {
        await _audioRecorder.stop();
      } catch (e) {
        print('Error stopping audio recorder: $e');
      }
    }
  }

  /// Add audio chunk from Frame microphone
  void addAudioChunk(Uint8List audioData) {
    _recognitionManager.addAudioChunk(audioData);
  }

  /// Perform song recognition
  Future<void> _performRecognition() async {
    final result = await _recognitionManager.recognize();

    if (result == null || !result.isValid) {
      // No song detected or error
      if (currentSong != null) {
        // Check if song has ended
        final position = _positionTracker.getCurrentPosition();
        if (position > currentSong!.duration + 5) {
          await _endSong();
        }
      }
      return;
    }

    // Got a recognition result
    await _handleRecognitionResult(result);
  }

  /// Handle song recognition result
  Future<void> _handleRecognitionResult(RecognitionResult result) async {
    // Check if this is a new song or continuation
    if (currentSong == null) {
      // New song detected
      await _startNewSong(result);
    } else {
      // Check if same song or different song
      if (result.title == currentSong!.title &&
          result.artist == currentSong!.artist) {
        // Same song - update position
        _updateSongPosition(result);
      } else {
        // Different song detected
        if (result.confidence > 0.75) {
          // High confidence for song switch
          await _switchSong(result);
        }
      }
    }
  }

  /// Start tracking a new song
  Future<void> _startNewSong(RecognitionResult result) async {
    currentSong = CurrentSong(
      title: result.title,
      artist: result.artist,
      album: result.album,
      duration: result.duration ?? 180.0, // Default 3 minutes
      detectedAt: DateTime.now(),
      hasLyrics: false,
      confidence: result.confidence,
    );

    // Start position tracking
    if (result.offsetSeconds != null) {
      _positionTracker.startSong(
        DateTime.now(),
        result.offsetSeconds!,
        result.apiLatency ?? 0,
      );
    }

    // Add to history
    _historyManager.addSong(currentSong!);

    // Set state to processing while we fetch lyrics
    appState = AppState.processing;

    // Fetch lyrics
    final lrcData = await _lyricsManager.fetchLyrics(currentSong!);

    if (lrcData != null && lrcData.isNotEmpty) {
      currentSong!.hasLyrics = true;
      currentSong!.lrcData = lrcData;
      appState = AppState.songDetectedWithLyrics;
    } else {
      appState = AppState.songDetectedNoLyrics;
    }
  }

  /// Update position tracking for current song
  void _updateSongPosition(RecognitionResult result) {
    if (result.offsetSeconds != null) {
      _positionTracker.recalibrate(
        result.offsetSeconds!,
        DateTime.now(),
        result.confidence,
        result.apiLatency ?? 0,
      );
    }
  }

  /// Switch to a different song
  Future<void> _switchSong(RecognitionResult result) async {
    await _endSong();
    await _startNewSong(result);
  }

  /// End current song tracking
  Future<void> _endSong() async {
    currentSong = null;
    appState = AppState.listening;
    _positionTracker.reset();
  }

  /// Update Frame display
  Future<void> _updateDisplay() async {
    // Get current position
    final position = _positionTracker.getCurrentPosition();

    // Get current and next lyrics chunks
    LyricsChunk? currentChunk;
    LyricsChunk? nextChunk;

    if (currentSong != null && currentSong!.hasLyrics) {
      currentChunk = _lyricsManager.getCurrentChunk(position);
      nextChunk = _lyricsManager.getNextChunk(position);
    }

    // Format display
    final displayText = _displayManager.formatDisplay(
      appState: appState,
      currentSong: currentSong,
      currentChunk: currentChunk,
      nextChunk: nextChunk,
      currentPosition: position,
    );

    // Send to Frame (if connected)
    if (frame != null && audioSource == AudioSource.frame) {
      try {
        await frame!.display.showText(displayText);
      } catch (e) {
        print('Display update error: $e');
      }
    }

    // Also send to app UI callback
    if (onDisplayUpdate != null) {
      onDisplayUpdate!(displayText);
    }
  }

  /// Get current app state (for UI)
  String get currentStateString {
    switch (appState) {
      case AppState.listening:
        return 'Listening for music...';
      case AppState.processing:
        return 'Processing song...';
      case AppState.songDetectedNoLyrics:
        return 'Playing: ${currentSong?.title ?? "Unknown"}';
      case AppState.songDetectedWithLyrics:
        return 'Playing with lyrics: ${currentSong?.title ?? "Unknown"}';
    }
  }

  /// Check if currently running
  bool get isRunning => _isRunning;

  /// Dispose resources
  void dispose() {
    _audioRecorder.dispose();
  }
}
